tests/test_cloudfiles_storage.py

```diff
@@ -0,0 +1,391 @@
+                    }
+                ],
+                "id": "v2.0",
+                "status": "CURRENT"
+            }
+        }).encode("utf8")]
+
+    def authentication_handler(
+            self, environ: "Environ", start_response: "StartResponse") -> List[bytes]:
+        body_size = int(environ.get("CONTENT_LENGTH", 0))
+        body = json.loads(environ["wsgi.input"].read(body_size))
+
+        # Forcing a 401 since swift service won't let us provide it
+        if self.keystone_credentials == {}:
+            start_response("401 Unauthorized", [("Content-type", "text/plain")])
+            return [b"Unauthorized keystone credentials."]
+        if not self._has_valid_credentials(body["auth"]["RAX-KSKEY:apiKeyCredentials"]):
+            start_response("403 Forbidden", [("Content-type", "text/plain")])
+            return [b"Invalid keystone credentials."]
+
+        start_response("200 OK", [("Content-type", "application/json")])
+        return [json.dumps({
+            "access": {
+                "serviceCatalog": [{
+                    "endpoints": [
+                        {
+                            "tenantId": "MOSSO-TENANT",
+                            "publicURL": self.swift_service.url("/v2.0/MOSSO-TENANT"),
+                            "internalURL": self.internal_cloudfiles_service.url(
+                                "/v2.0/MOSSO-TENANT"),
+                            "region": "DFW"
+                        },
+                        {
+                            "tenantId": "MOSSO-TENANT",
+                            "publicURL": self.alt_cloudfiles_service.url("/v2.0/MOSSO-TENANT"),
+                            "internalURL": self.alt_cloudfiles_service.url("/v2.0/MOSSO-TENANT"),
+                            "region": "ORD"
+                        }
+                    ],
+                    "name": "cloudfiles",
+                    "type": "object-store"
+                }],
+                "user": {
+                    "RAX-AUTH:defaultRegion": "DFW",
+                    "roles": [{
+                        "name": "object-store:default",
+                        "tenantId": "MOSSO-TENANT",
+                        "id": "ID"
+                    }],
+                    "name": "USER",
+                    "id": "IDENTIFIER"
+                },
+                "token": {
+                    "expires": "2019-07-18T05:47:13.090Z",
+                    "RAX-AUTH:authenticatedBy": ["APIKEY"],
+                    "id": "KEY",
+                    "tenant": {
+                        "name": "MOSSO-TENANT",
+                        "id": "MOSSO-TENANT"
+                    }
+                }
+            }
+        }).encode("utf8")]
+
+    def object_head_account_handler(
+            self, environ: "Environ", start_response: "StartResponse") -> List[bytes]:
+        start_response("204 OK", [
+            ("Content-type", "text-plain"),
+            ("temp-url-key", "TEMPKEY")
```

Comment:
Note that the actual header returned is something like `X-Account-Meta-Temp-Url-Key` (I'm not 100% certain on the capitalization of `URL` in the header).

Reply:
Headers are case-insensitive by the spec so it shouldn't matter what we return (although it's good to return a *different* case than we check in the code to verify that the HTTP client is doing "the right thing"), however to your point this is definitely the incorrect mock response. (The `swiftclient.head_account()` call automatically strips off any `X-Account-Meta-` prefix before returning it, at least in my actual tests against `Cloudfiles`, so the actual implementation is correct.)

Reply:
While headers are, conceptually, meant to be case-insensitive, they are sent "over the wire" with some casing. The problem we ran into with Pyrax (and the whole need for this test and the corresponding code to begin with) is that `requests` changed their interface, so that doing a *lookup* to the header would ignore case (i.e. `headers['some-HeADER']` would return the value of the header spelled `Some-Header` over the wire), but they changed it so that *iterating* over the headers returned the case as sent over the wire (i.e. you'd get a string `'some-HeADER'`. Then, when Pyrax tried to "search" for the header itself, it couldn't find it using newer versions of `requests`.

Anyway, making the test return what Cloudfiles returns should validate whether the code is working or not.